readme.md
[![NPM](https://img.shields.io/npm/v/graygelf.svg)](https://npmjs.com/package/graygelf) [![Build Status](https://secure.travis-ci.org/wavded/graygelf.svg)](http://travis-ci.org/wavded/graygelf) [![Coverage Status](https://coveralls.io/repos/github/wavded/graygelf/badge.svg?branch=master)](https://coveralls.io/github/wavded/graygelf?branch=master) [![code-style](https://img.shields.io/badge/code%20style-adc-brightgreen.svg?style=flat)](https://github.com/applieddataconsultants/eslint-config-adc)

GrayLog2 GELF UDP logging, streaming, chunking, and more. Production hardened. Includes client and server implementations. AFAIK a complete [GELF](http://graylog2.org/gelf#specs) implementation.

![GrayGelf](https://raw.github.com/wavded/graygelf/master/graygelf.png)

# Install

```
npm install graygelf
```

# Example

```javascript
var log = require('graygelf')('graylog.server.local')
log.on('message', console.log) // output messages to console

// setup global custom fields to be passed with every message
log.fields.facility = 'redicomps'

// printf style "hello world"
log.info('hello %s', 'world')

// concat by space style "hello world"
log.info('hello', 'world')

// stays context aware
redis.on('error', log.error)

// include a full message and custom fields using .a(ttach)
log.info.a('short', 'full', { foo: 'bar' })
log.info.a('short', 'full', { foo: 'bar' })

// if an Error is passed as the only argument...
var er = new Error('oh no!')
log.info(er)
// ... it expands to:
log.info.a(er.message, er.stack)

// writable streams can be created
var infostream = log.stream('info')
var rstream = require('fs').createReadStream(__filename)
rstream.pipe(infostream) // lines automatically split up and sent seperately

// raw gelf: version, host, and timestamp will be supplied if missing
log.raw({
  // version: '1.1',
  // host: 'wavded',
  short_message: 'oh no!',
  full_message: 'howdy',
  // timestamp: 1412087767.704356,
  level: 6,
  _foo: 'bar'
})
```

# Setup

By `host` string (uses defaults below for other options):

```javascript
var log = require('graygelf')('graylog.server.local')
```

By `options` object:

```javascript
var log = require('graygelf')({
  host: 'graylog.server.local',
  port: 23923
})
```

Available `options` are:

```
host
  - graylog host (default: 'localhost')
port
  - graylog port (default: 12201)
chunkSize
  - size of chunked messages in bytes (default: 1240)
compressType
  - compression 'gzip' or 'deflate' (default: 'deflate')
alwaysCompress
  - whether to always compress or go by chunkSize (default: false)
mock
  - don't send messages to GrayLog2 (default: false)
```

# API

## event: error

Emits errors that may occur while parsing and sending GELF messages.

## event: message

Emits GELF JSON messages that will be send over UDP. Useful for redirecting output to stdout in development.

```javascript
log.on('message', function (gelf) {
  console.log(gelf.level, gelf.short_message, gelf.long_message)
})
```

## log.fields

Add global custom fields to be included in every message. Custom fields allow you to more interesting searches and sorting inside GrayLog2 servers.

```javascript
log.fields.facility = 'facility'
```

Note: `fields` is plain JavaScript object.

## log{level}(message)

GrayGelf maps the syslog levels to functions. All functions have the same semantics as `console.log` (i.e. [printf style](http://nodejs.org/api/util.html#util_util_format_format)):

```javascript
log.emerg('oh %s', 's*#t')              // 0 - alias: panic
log.alert('act', 'immediately')         // 1
log.crit('act %j', [ 'really soon' ])   // 2
log.error('expected %d, got %d', 1, 5)  // 3 - alias: err
log.warn('take note, it may bite')      // 4 - alias: warning
log.notice('unusual %s', 'behavior')    // 5
log.info('hello', 'world')              // 6
log.debug('value is', a)                // 7
```

## log{level}.a(short, long, custom)

There also is an `a(ttach)` method to include a full message.

```javascript
log.crit.a('short message', 'full message')
```

The `a(ttach)` method can have an optional third argument to define custom fields that will be passed to Graylog2.

```javascript
log.info.a('short message', 'full message', { custom: 'field' })
```

## log.stream(level)

Create a writable stream to pipe log messages into:

```javascript
var stream = log.stream('info')
```

Streams automatically break lines up and pass each line to GrayLog2 at the specified level.

## log.raw(gelf)

Pass a raw [GELF](http://www.graylog2.org/resources/gelf/specification) message. The following fields will be populated if absent: `version`, `host`, and `timestamp`.

```javascript
log.raw({
  version: '1.1',
  host: 'wavded',
  short_message: 'oh no!',
  full_message: 'howdy',
  timestamp: 1412087767.704356,
  level: 6,
  _foo: 'custom field'
})
```

Note: No global custom fields (`log.fields`) are included when using `log.raw`.

# Server

Make your own GrayLog UDP server or proxy messages to GrayLog. A GrayGelf server handles `zlib`, `gzip` and GELF chunked messages.

## Example

```javascript
var gelfserver = require('graygelf/server')
var server = gelfserver()
server.on('message', function (gelf) {
  // handle parsed gelf json
  console.log('received message', gelf.short_message)
})
server.listen(12201)
```

## event: message

Emits parsed GELF JSON messages.

## event: data

Emits raw GELF buffers (useful for proxying).

## event: error

Emits errors captured from udp or parsing.

## server.listen(port = 12201, address = "0.0.0.0")

Start listening on a port and bind address. Both parameters are optional. Defaults to typical GrayLog2 server defaults.

## server.close()

Close down a server and stop receiving messages.

## server.unref()

Allow the Node process to terminate if the server is the only thing keeping it alive.

## server.pipe(client)

```javascript
var server = require('graygelf/server')().listen()
var client = require('graygelf')('proxy-dest.graylog.local')

server.pipe(client) // establish proxy (straight UDP transfer)
```


changelog.md
# 2.0.0 / 2018-04-12

- Drop support for Node 6 and 8

# 1.2.0 / 2018-04-12

- Add option to always compress messages [ndreckshage]

# 1.1.0 / 2017-10-23

- Added support for non-compressed GELF messages in client and server [shanewholloway]
- Added package.json fields for other people to follow from npmjs to github repository [shanewholloway]

# 1.0.2 / 2016-12-24

- Increase to millisecond precision in timestamp fallback function [rho24]

# 1.0.0 / 2014-09-30

- THIS IS A BREAKING RELEASE. Please consult the updated documentation.

- add; new simplier client and server apis

- add; support for console.log style client logging

- add; support for raw gelf messages from client

- add; support for Error objects

- add; support for global custom fields

- add; clearer api for per-message custom fields

- add; line by line parsing of streams for client

- add; unref() for servers

# 0.6.0 / 2014-09-29

- add; auto unref client

# 0.5.1 / 2014-08-26

- fix; bad microtime timestamp
- mod; spec link in readme

# 0.5.0 / 2014-08-20

- add; console.log style message concat
- add; optional microtime dep
- mod; make facility optional (gelf 1.1)

# 0.4.0 / 2014-03-18

- drop; node 0.8 support
- add; mocking for client
- add; bound logging level functions

# 0.3.7 / 2014-01-08

- updated; deps
- fix; do not write logs if client is closed
- added; npm badge
- updated; tests for node 0.10.0

# 0.3.6 / 2013-03-07

- fixed; undefined level emitted when .log is used [danmilon]

# 0.3.5 / 2013-02-19

- updated; do not get the hostname for every message [danmilon]
- added; custom log method [danmilon]

# 0.3.4 / 2013-02-05

- fixed; handle out of order chunked message properly
- fixed; catch duplicate ids errors in server

# 0.3.3 / 2013-02-04

- updated; use crypto.randomBytes for id

# 0.3.2 / 2013-01-24

- updated; udp error handling
- updated; client 'message' includes level text instead of number
- fixed; don't pass full_message if none
- added; more tests

# 0.3.1 / 2012-11-01

- added; message event for client
- added; link to GELF docs
- added; graygelf img

# 0.3.0 / 2012-10-05

- added; proxy server support

# 0.2.1 / 2012-10-05

- updated; desc and keywords
- removed; node 0.6 support
- added; .npmignore

# 0.2.0 / 2012-10-05

- added; more tests
- added; close methods for client and server
- added; compressType option to client

# 0.1.0 / 2012-10-04

- added; graygelf server support, changed api


lib/client.js
var zlib   = require('zlib')
var dgram  = require('dgram')
var os     = require('os')
var crypto = require('crypto')
var Stream = require('stream')
var format = require('util').format
var split  = require('split')

/**
 * Prefer microtime accuracy but fallback if unable to use
 */
var timestamp
try       { timestamp = require('microtime').nowDouble }
catch (e) { timestamp = function () { return Date.now() / 1000 >> 0 } }

/**
 * Available logging levels
 * @type {Array}
 */
var LOG_LEVELS = {
  'emerg'   : 0,
  'panic'   : 0,
  'alert'   : 1,
  'crit'    : 2,
  'error'   : 3,
  'err'     : 3,
  'warn'    : 4,
  'warning' : 4,
  'notice'  : 5,
  'info'    : 6,
  'debug'   : 7
}

var GrayGelf = function (opts) {
  if (!(this instanceof GrayGelf)) {
    return new GrayGelf(opts)
  }

  if (typeof opts === 'string') opts = { host: opts }
  if (!opts) opts = {}

  this.graylogHost  = opts.host || 'localhost'
  this.graylogPort  = opts.port || 12201
  this.fields       = {}

  this.chunkSize    = opts.chunkSize || GrayGelf.CHUNK_WAN
  this.compressType = (opts.compressType || '') === 'gzip' ? 'gzip' : 'deflate'
  this.hostname     = os.hostname()

  if (!opts.mock) {
    this._udp = dgram.createSocket('udp4')
    this._udp.on('error', this._emitError.bind(this))
    this._udp.unref()
  }

  this.writable = true // writable stream
  this._setupLevels()
}

GrayGelf.prototype = Object.create(Stream.prototype)

GrayGelf.prototype._setupLevels = function () {
  var graygelf = this

  Object.keys(LOG_LEVELS).forEach(function (name) {
    var level = LOG_LEVELS[name]

    graygelf[name] = function () {
      var gelf
      if (arguments.length === 1 && arguments[0] instanceof Error) {
        gelf = graygelf._prepGelf(level, arguments[0].message, arguments[0].stack)
      }
      else {
        gelf = graygelf._prepGelf(level, format.apply(null, arguments))
      }
      graygelf._send(gelf)
      return gelf
    }

    graygelf[name].a = function (short, long, fields) {
      var gelf = graygelf._prepGelf(level, short, long, fields)
      graygelf._send(gelf)
      return gelf
    }
  })
}

GrayGelf.prototype.write = function (chunk) {
  if (!this._udp) return
  this._udp.send(chunk, 0, chunk.length, this.graylogPort, this.graylogHost)
}

GrayGelf.prototype._emitError = function (er) { this.emit('error', er) }

GrayGelf.prototype._prepGelf = function (level, short, long, fields) {
  var gelf = {
    version: '1.1',
    host: this.hostname,
    short_message: short,
    timestamp: timestamp(),
    level: level
  }

  if (!fields) fields = {}
  if (long) gelf.full_message = long

  for (var i in this.fields) gelf['_'+i] = this.fields[i]
  for (var j in fields) gelf['_'+j] = fields[j]

  this.emit('message', gelf)
  return gelf
}

GrayGelf.prototype._send = function (gelf) {
  var gelfbuf = new Buffer(JSON.stringify(gelf))
  var graygelf = this

  zlib[this.compressType](gelfbuf, function (er, message) {
    /* istanbul ignore if */
    if (er) return graygelf.emit('error', er)

    if (message.length > graygelf.chunkSize) {
      var total = Math.ceil(message.length / graygelf.chunkSize)
      var offset = 0

      crypto.randomBytes(6, function (er, idBuf) {
        /* istanbul ignore if */
        if (er) return graygelf.emit('error', er)

        for (var i = 0; i < total; i++) {
          var bytesToSend = offset + graygelf.chunkSize < message.length ? graygelf.chunkSize : message.length - offset
          var chunk = new Buffer(bytesToSend + 12)
          chunk[0] = 0x1e
          chunk[1] = 0x0f
          idBuf.copy(chunk, 2, 0, 6)
          chunk[10] = i
          chunk[11] = total

          message.copy(chunk, 12, offset, offset+bytesToSend)
          offset += bytesToSend
          graygelf.write(chunk)
        }
      })
    }
    else graygelf.write(message)
  })
}

GrayGelf.prototype.stream = function (name) {
  if (!(name in LOG_LEVELS)) throw new Error('invalid stream name')
  var stream = new Stream()
  var lines = split()
  stream.writable = true
  stream.write = this[name].bind(this)
  stream.end = function(){} // 'end' event is noop, stream always stays open
  lines.pipe(stream)
  return lines
}

GrayGelf.prototype.raw = function (raw) {
  if (!raw.version) raw.version = '1.1'
  if (!raw.host) raw.host = this.hostname
  if (!raw.timestamp) raw.timestamp = timestamp()
  this._send(raw)
  return raw
}

GrayGelf.CHUNK_WAN  = 1240
GrayGelf.CHUNK_LAN  = 8154
GrayGelf.LOG_LEVELS = LOG_LEVELS

module.exports = GrayGelf